File: rust/src/advent_of_code12.rs
use crate::advent_of_code4::Matrix;
;


pub fn advent_of_code12a(mut input_str: &str) -> i32 {
    
    let mut ncols: usize = 0;
    let mut nrows: usize = 0;

    for line in input_str.lines() {
        println!("{}", line);
        nrows = line.len();
        ncols += 1;
    }

    let mut puzzle_matrix = Matrix::new(nrows, ncols, ' ');

    for line in input_str.lines() {
        let mut col_idx = 0;
        for c in line.chars() {
            matrix.set(row_idx, col_idx, c);
            col_idx += 1;
        }
        row_idx += 1;

    println!("nrows: {}, ncols: {}", nrows, ncols);
    println!()
}

File: python/src/advent_of_code/2024/advent_of_code12.py
from dataclasses import dataclass
from python.src.advent_of_code.utils import Matrix

@dataclass
class Region:
    area: int
    perimeter: int


def advent_of_code12a(input_str: str) -> int:

    input_str = input_str.split()
    input_str = [list(line) for line in input_str]


    nrows = len(input_str)
    ncols = len(input_str[0])

    input_matrix = Matrix(nrows, ncols)
    
    for row in range(0, nrows):
        for col in range(0, ncols):
            input_matrix[row, col] = input_str[row][col]

    input_matrix.print_matrix()

    return 0

if __name__=="__main__":
    
    test_input1 = """AAAA
BBCD
BBCC
EEEC"""

    test_result1 = advent_of_code12a(test_input1)
    assert test_result1==772

    test_input2 = """RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE"""

    test_result2 = advent_of_code12a(test_input2)
    assert test_result2==1930

    with open("../data/advent_of_code12/puzzle_input.txt", 'r') as f:
        puzzle_input = f.read()

    puzzle_output = advent_of_code12a(puzzle_input)
    print(f"puzzle_output is: {puzzle_output}")
    